core/src/services/persy/backend.rs

```diff
@@ -0,0 +1,245 @@
+            .open(&datafile_path)
+            .map_err(|e| {
+                Error::new(ErrorKind::ConfigInvalid, "open db")
+                    .with_context("service", Scheme::Persy)
+                    .with_context("datafile", datafile_path.clone())
+                    .set_source(e)
+            })?;
+
+        // This function will only be called on database creation
+        fn init(
+            persy: &persy::Persy,
+            segment_name: &str,
+            index_name: &str,
+        ) -> std::result::Result<(), Box<dyn std::error::Error>> {
+            let mut tx = persy.begin()?;
+
+            if !tx.exists_segment(segment_name)? {
+                tx.create_segment(segment_name)?;
+            }
+            if !tx.exists_index(index_name)? {
+                tx.create_index::<String, persy::PersyId>(index_name, persy::ValueMode::Replace)?;
+            }
+
+            let prepared = tx.prepare()?;
+            prepared.commit()?;
+
+            println!("Segment and Index successfully created");
+            Ok(())
+        }
+
+        Ok(PersyBackend::new(Adapter {
+            datafile: datafile_path,
+            segment,
+            index,
+            persy,
+        }))
+    }
+}
+
+/// Backend for persy services.
+pub type PersyBackend = kv::Backend<Adapter>;
+
+#[derive(Clone)]
+pub struct Adapter {
+    datafile: String,
+    segment: String,
+    index: String,
+    persy: persy::Persy,
+}
+
+impl Debug for Adapter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Adapter");
+        ds.field("path", &self.datafile);
+        ds.field("segment", &self.segment);
+        ds.field("index", &self.index);
+        ds.finish()
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for Adapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Persy,
+            &self.datafile,
+            Capability {
+                read: true,
+                write: true,
+                delete: true,
+                blocking: true,
+                ..Default::default()
+            },
+        )
+    }
+
+    async fn get(&self, path: &str) -> Result<Option<Vec<u8>>> {
+        self.blocking_get(path)
+    }
+
+    fn blocking_get(&self, path: &str) -> Result<Option<Vec<u8>>> {
+        let mut read_id = self
+            .persy
+            .get::<String, persy::PersyId>(&self.index, &path.to_string())
+            .map_err(parse_error)?;
+        if let Some(id) = read_id.next() {
+            let value = self.persy.read(&self.segment, &id).map_err(parse_error)?;
+            return Ok(value);
+        }
+
+        Ok(None)
+    }
+
+    async fn set(&self, path: &str, value: &[u8]) -> Result<()> {
+        self.blocking_set(path, value)
+    }
+
+    fn blocking_set(&self, path: &str, value: &[u8]) -> Result<()> {
+        let mut tx = self.persy.begin().map_err(parse_error)?;
```

Comment:
We don't need `tx` to read data?

Reply:
https://persy.rs/posts/getting-started.html

```rust
   let mut tx = persy.begin()?;
    tx.create_segment("data")?;
    tx.create_index::<String, PersyId>("index", ValueMode::Replace)?;
    let prepared = tx.prepare()?;
    prepared.commit()?;

    let mut tx = persy.begin()?;
    let rec = "record content".as_bytes();
    let id = tx.insert("data", rec)?;

    tx.put::<String, PersyId>("index", "key".to_string(), id)?;
    let prepared = tx.prepare()?;
    prepared.commit()?;

    let read_id = persy.get::<String, PersyId>("index", &"key".to_string())?;
    if let Some(id) = read_id.next() {
        let value = persy.read("data", &id)?;
        assert_eq!(Some(rec.to_vec()), value);
    }
```

Reply:
According to the example, we do not need to use "tx" when reading data.